dns_sd_windows.c

```diff
@@ -0,0 +1,257 @@
+				struct sockaddr_in6* saddr = (struct sockaddr_in6*)unicast->Address.lpSockaddr;
+				static const unsigned char localhost[] = { 0, 0, 0, 0, 0, 0, 0, 0,
+														  0, 0, 0, 0, 0, 0, 0, 1 };
+				static const unsigned char localhost_mapped[] = { 0, 0, 0,    0,    0,    0, 0, 0,
+																 0, 0, 0xff, 0xff, 0x7f, 0, 0, 1 };
+				if ((unicast->DadState == NldsPreferred) &&
+					memcmp(saddr->sin6_addr.s6_addr, localhost, 16) &&
+					memcmp(saddr->sin6_addr.s6_addr, localhost_mapped, 16)) {
+					if (num_sockets < max_sockets) {
+						int sock = mdns_socket_open_ipv6(saddr);
+						if (sock >= 0) {
+							sockets[num_sockets++] = sock;
+						}
+					}
+				}
+			}
+#endif
+		}
+	}
+
+	free(adapter_address);
+
+	for (int isock = 0; isock < num_sockets; ++isock) {
+		unsigned long param = 1;
+		ioctlsocket(sockets[isock], FIONBIO, &param);
+	}
+
+	return num_sockets;
+}
+
+
+static int
+query_callback(int sock, const struct sockaddr* from, size_t addrlen,
+	mdns_entry_type_t entry, uint16_t transaction_id,
+	uint16_t rtype, uint16_t rclass, uint32_t ttl,
+	const void* data, size_t size, size_t offset, size_t length,
+	void* user_data) {
+
+
+	char addrbuffer[64];
+	char servicebuffer[64];
+	char namebuffer[256];
+
+	struct dns_sd_discovery_data* dd = (struct dns_sd_discovery_data*)user_data;
+	if (dd == NULL) {
+		IIO_ERROR("DNS SD: Missing info structure. Stop browsing.\n");
+		goto quit;
+	}
+
+	if (rtype != MDNS_RECORDTYPE_SRV)
+		goto quit;
+
+	getnameinfo((const struct sockaddr*)from, (socklen_t)addrlen,
+		addrbuffer, NI_MAXHOST, servicebuffer, NI_MAXSERV,
+		NI_NUMERICSERV | NI_NUMERICHOST);
+
+	mdns_record_srv_t srv = mdns_record_parse_srv(data, size, offset, length,
+		namebuffer, sizeof(namebuffer));
+	IIO_DEBUG("%s : SRV %.*s priority %d weight %d port %d\n",
+		addrbuffer,
+		MDNS_STRING_FORMAT(srv.name), srv.priority, srv.weight, srv.port);
+
+	// Go to the last element in the list
+	while (dd->next != NULL)
+		dd = dd->next;
+
+	if (srv.name.length > 1)
+	{
+		dd->hostname = malloc(srv.name.length);
+		iio_strlcpy(dd->hostname, srv.name.str, srv.name.length);
+		dd->hostname[srv.name.length - 1] = 0;
+	}
+	iio_strlcpy(dd->addr_str, addrbuffer, strlen(addrbuffer));
+	dd->port = srv.port;
+
+	IIO_DEBUG("DNS SD: added %s (%s:%d)\n", dd->hostname, dd->addr_str, dd->port);
+	// A list entry was filled, prepare new item on the list.
+	if (new_discovery_data(&dd->next)) {
+		IIO_ERROR("DNS SD mDNS Resolver : memory failure\n");
+	}
+
+quit:
+	return 0;
+}
+
+int dnssd_find_hosts(struct dns_sd_discovery_data** ddata)
+{
+
+	WORD versionWanted = MAKEWORD(1, 1);
+	WSADATA wsaData;
+	if (WSAStartup(versionWanted, &wsaData)) {
+		printf("Failed to initialize WinSock\n");
+		return -1;
+	}
+
+	struct dns_sd_discovery_data* d;
+
+	IIO_DEBUG("DNS SD: Start service discovery.\n");
+
+	if (new_discovery_data(&d) < 0) {
+		return -ENOMEM;
+	}
+	*ddata = d;
+
+	size_t capacity = 2048;
+	void* buffer = malloc(capacity);
```

Comment:
buffer is never checked to see if malloc() fails

Reply:
fixed